tsl/src/chunk_copy.c

```diff
@@ -0,0 +1,581 @@
+struct ChunkCopy
+{
+	/* catalog data */
+	FormData_chunk_copy_activity fd;
+	/* current stage being executed */
+	const ChunkCopyStage *stage;
+	/* chunk to copy */
+	Chunk *chunk;
+	/* from/to foreign servers */
+	ForeignServer *src_server;
+	ForeignServer *dst_server;
+	/* temporary memory context */
+	MemoryContext mcxt;
+};
+
+static HeapTuple
+chunk_copy_activity_make_tuple(const FormData_chunk_copy_activity *fd, TupleDesc desc)
+{
+	Datum values[Natts_chunk_copy_activity];
+	bool nulls[Natts_chunk_copy_activity] = { false };
+	memset(values, 0, sizeof(values));
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_id)] = Int32GetDatum(fd->id);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_operation_id)] =
+		NameGetDatum(&fd->operation_id);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_backend_pid)] =
+		Int32GetDatum(fd->backend_pid);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_completed_stage)] =
+		NameGetDatum(&fd->completed_stage);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_time_start)] =
+		TimestampTzGetDatum(fd->time_start);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_chunk_id)] =
+		Int32GetDatum(fd->chunk_id);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_source_node_name)] =
+		NameGetDatum(&fd->source_node_name);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_dest_node_name)] =
+		NameGetDatum(&fd->dest_node_name);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_delete_on_src_node)] =
+		BoolGetDatum(fd->delete_on_src_node);
+	return heap_form_tuple(desc, values, nulls);
+}
+
+static void
+chunk_copy_activity_insert_rel(Relation rel, const FormData_chunk_copy_activity *fd)
+{
+	CatalogSecurityContext sec_ctx;
+	HeapTuple new_tuple;
+
+	new_tuple = chunk_copy_activity_make_tuple(fd, RelationGetDescr(rel));
+
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+	ts_catalog_insert(rel, new_tuple);
+	ts_catalog_restore_user(&sec_ctx);
+	heap_freetuple(new_tuple);
+}
+
+static void
+chunk_copy_activity_insert(const FormData_chunk_copy_activity *fd)
+{
+	Catalog *catalog;
+	Relation rel;
+
+	catalog = ts_catalog_get();
+	rel = table_open(catalog_get_table_id(catalog, CHUNK_COPY_ACTIVITY), RowExclusiveLock);
+
+	chunk_copy_activity_insert_rel(rel, fd);
+	table_close(rel, RowExclusiveLock);
+}
+
+static ScanTupleResult
+chunk_copy_activity_tuple_update(TupleInfo *ti, void *data)
+{
+	ChunkCopy *cc = data;
+	Datum values[Natts_chunk_copy_activity];
+	bool nulls[Natts_chunk_copy_activity];
+	CatalogSecurityContext sec_ctx;
+	bool should_free;
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+	HeapTuple new_tuple;
+
+	heap_deform_tuple(tuple, ts_scanner_get_tupledesc(ti), values, nulls);
+
+	/* We only update the "completed_stage" field */
+	Assert(NULL != cc->stage);
+	values[AttrNumberGetAttrOffset(Anum_chunk_copy_activity_completed_stage)] =
+		CStringGetDatum(cc->stage->name);
```

Comment:
Note that this is now Name type. In other places we `DirectFunctionaCall1(namein, name)` to construct a name type. Alternatively, create a `NameData` on the stack and `namestrcpy()` the name into it followed by `NameGetDatum`

Reply:
`NameGetDatum` translates to the `CStringGetDatum` in the macro if I am not mistaken

Reply:
It's `#define NameGetDatum(X) CStringGetDatum(NameStr(*(X)))` from:
```
/*
 * Representation of a Name: effectively just a C string, but null-padded to
 * exactly NAMEDATALEN bytes.  The use of a struct is historical.
 */
typedef struct nameData
{
	char		data[NAMEDATALEN];
} NameData;
typedef NameData *Name;

#define NameStr(name)	((name).data)
```
A recent bugfix in master was so that we don't get memory overflows by not allocating enough `char` entries. See [here](https://github.com/timescale/timescaledb/pull/3317#issuecomment-858459915)

Reply:
I guess the problem was that the length of the string pointer for name type was larger then `NAMEDATALEN`?

Reply:
The opposite actually, it would `memcpy()` from the source which didn't have `NAMEDATALEN` characters allocated because it was a C string, and it would do the memory overflow.

Reply:
Ok cool, using `namein` function as suggested

Reply:
> The opposite actually, it would `memcpy()` from the source which didn't have `NAMEDATALEN` characters allocated because it was a C string, and it would do the memory overflow.

This is exactly what I meant to say :) good to know